### PR TITLE
fix: A referenced default package must be a plugin package

### LINKED ISSUE #152037: `default_package` references now create an error if they reference a package that isn't a plugin implementation
### Steps to reproduce

TLDR; try to add/compile an app with a plugin using `default_package` to endorse a package.

## Repro steps

1. Clone sample app https://github.com/Equartey/flutter_default_plugin_repro
2. Checkout Flutter master channel, `flutter channel master`
   1. Error is present at least between `>=3.24.0-1.0.pre.151 >3.24.0-1.0.pre.180`
3. Run `$ flutter pub get`.
4. Error should be present. App will also fail to compile.
   1. Alternative: switch to Flutter stable channel and dependencies resolve along with the app compiling.

### Expected results

Flutter should be able to resolve the endorsed package like in previous versions. Or if this is intended, documentation should be updated to reflect the new behavior. 

### Actual results

The following errors prevent the plugin from being fetched and/or the app from compiling.

```
Package amplify_secure_storage:linux references amplify_secure_storage_dart:linux as the default plugin, but the
package does not exist.
Ask the maintainers of amplify_secure_storage to either avoid referencing a default implementation via `platforms:
linux: default_package: amplify_secure_storage_dart` or create a plugin named amplify_secure_storage_dart.


Package amplify_secure_storage:windows references amplify_secure_storage_dart:windows as the default plugin, but the
package does not exist.
Ask the maintainers of amplify_secure_storage to either avoid referencing a default implementation via `platforms:
windows: default_package: amplify_secure_storage_dart` or create a plugin named amplify_secure_storage_dart.
```

### Code sample

[Sample Repo](https://github.com/Equartey/flutter_default_plugin_repro)

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details open><summary>Logs</summary>

https://pastebin.com/fj7KjACU

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel master, 3.24.0-1.0.pre.180, on macOS 14.5 23F79 darwin-arm64, locale en-US)
    • Flutter version 3.24.0-1.0.pre.180 on channel master at /Users/quaelija/fvm/versions/stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 36dac294e9 (25 hours ago), 2024-07-18 12:07:26 -0400
    • Engine revision d58ba74250
    • Dart version 3.6.0 (build 3.6.0-54.0.dev)
    • DevTools version 2.37.1

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/quaelija/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/quaelija/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15F31d
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.7+0-17.0.7b1000.6-10550314)

[✓] VS Code (version 1.91.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.92.0

[✓] Connected device (7 available)
    • Android SDK built for arm64 (mobile) • emulator-5554                        • android-arm64  • Android 7.1.1
      (API 25) (emulator)
    • iphone (mobile)                      • 00008120-0014189C017BC01E            • ios            • iOS 17.5.1 21F90
    • iPhone 15 Pro (mobile)               • 9769DBEE-D114-4A35-B4BF-360F06E79C83 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-17-4 (simulator)
    • iPhone 15 Pro Max (mobile)           • B66B1944-2468-4C2E-A646-67365BBDA3DC • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-17-0 (simulator)
    • macOS (desktop)                      • macos                                • darwin-arm64   • macOS 14.5 23F79
      darwin-arm64
    • Mac Designed for iPad (desktop)      • mac-designed-for-ipad                • darwin         • macOS 14.5 23F79
      darwin-arm64
    • Chrome (web)                         • chrome                               • web-javascript • Google Chrome
      126.0.6478.183

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by Gustl22**
Clarify that a referenced default package must also be a plugin package (one, which provides a pluginClass or a dartPluginClass).

Fixes #152037
More precisely https://github.com/flutter/flutter/issues/152037#issuecomment-2239909463.
Introduced in #137040

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Comment by Gustl22** [comment]
Open for review @stuartmorgan

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]
Thanks!

File: packages/flutter_tools/lib/src/flutter_plugins.dart [hunk 1]
```diff
@@ -1263,7 +1263,7 @@ Map<String, List<Plugin>> _resolvePluginImplementations(
         }
       } else {
         globals.printWarning(
-          'Package ${plugin.name}:$platformKey references $defaultImplPluginName:$platformKey as the default plugin, but the package does not exist.\n'
+          'Package ${plugin.name}:$platformKey references $defaultImplPluginName:$platformKey as the default plugin, but the package does not exist, or is not a plugin package.\n'
```
[the remark this answers is not in the document]
Comment (Gustl22) [reply packages/flutter_tools/lib/src/flutter_plugins.dart:1266]: This check involves all plugins, which have [this](https://github.com/flutter/flutter/blob/0b41781c780a2782dd3c2a39bc7e2cf1a0b16e70/packages/flutter_tools/lib/src/flutter_plugins.dart#L62) condition. So it's pre-platform-filtering. Means, all with the `flutter: plugin:` in the pubspec are considered. So I think the error message is correct that way.

We test the specific platform implementation a few lines above, which should cover your mentioned case (if I got it right): https://github.com/flutter/flutter/pull/152134/files#diff-68361f6fc9aafa5fa7de042eec4603dd4f321eb783a00edd57780ef78c1094bbR1259

File: packages/flutter_tools/lib/src/flutter_plugins.dart [hunk 2]
```diff
@@ -1263,7 +1263,7 @@ Map<String, List<Plugin>> _resolvePluginImplementations(
...
           'Ask the maintainers of ${plugin.name} to either avoid referencing a default implementation via `platforms: $platformKey: default_package: $defaultImplPluginName` '
           'or create a plugin named $defaultImplPluginName.\n',
```
[the remark this answers is not in the document]
Comment (Gustl22) [reply packages/flutter_tools/lib/src/flutter_plugins.dart:1268]: See the [other](https://github.com/flutter/flutter/pull/152134#discussion_r1699114020) comment.

File: packages/flutter_tools/lib/src/flutter_plugins.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (stuartmorgan) [reply packages/flutter_tools/lib/src/flutter_plugins.dart:1266]: Makes sense, thanks. Ideally the check should be stricter and require a platform implementation, but we can revisit that if it ever actually matters in practice.

**Review by stuartmorgan-g — APPROVED** [review]
LGTM

**Review by christopherfujino — APPROVED** [review]
LGTM

**Pull request merged**